File: LeetCodePython/py12.py
class Solution:
    #解法二：进制处理。最大优先
    def intToRoman(self, num: int) -> str:
        lookup = {
            1:'I',
            4:'IV',
            5:'V',
            9:'IX',
            10:'X',
            40:'XL',
            50:'L',
            90:'XC',
            100:'C',
            400:'CD',
            500:'D',
            900:'CM',
            1000:'M'  
        }
        result = ""
        for key in sorted(lookup.keys())[::-1]:
            items = num//key
            if items == 0:
                continue
            result += lookup[key]*items
            num -= items * key
        return result


class Solution1:
    # 解法一：暴力动态规划，以num数值建立数组
    def intToRoman(self, num: int) -> str:
        value = [1,4,5,9,10,40,50,90,100,400,500,900,1000]
        label = ['I','IV','V','IX','X','XL','L', 'XC', 'C','CD','D','CM', 'M']
        dp = [False] * (num + 1)
        dp[0] = True
        result = [''] * (num + 1)
        for i in range(num+1):
            for j in range(len(value)-1, -1, -1):
                if dp[i] and i+value[j]< num+1 and not dp[i+value[j]]:
                    dp[i+value[j]] = True
                    result[i+value[j]] = label[j] + result[i]
        # for i in range(num):
        #     print(dp[i], result[i])
        return result[num]
        
       
        
        
        

File: LeetCodeCPlusPlus/py12.cpp
#include<string>
#include<vector>
using namespace std;
// 整数转罗马数字，按从到小，依次转换，拼接
class Solution {
public:
    string intToRoman(int num) {
        vector<int> value = { 1000,900,500,400,100,90,50,40,10,9,5,4,1};
		vector<string> dic = { "M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I" };
        string result = "";
        int index = 0;
        while (num>0 && index<value.size()) {
            if (num>=value[index]) {
                result += dic[index];
                num -= value[index];
            }else
                index++;
        }
        // for(int i=0;i<value.size();i++) {
        //      if (num/value[i] == 0)
        //          continue;
        //      for (int j=0;j<num/value[i];j++)
        //          result += dic[i];
        //      num -= num/value[i]*value[i];
        // }
        return result;
    }
};